### ISSUE TITLE
Error Handling of Non-Ints/Floats for postion of xticks

### ISSUE THREAD

**Issue by gurkpet** [matplotlib/matplotlib#10147]
This appears to be the underlying cause of ticket #10130 

I think this is an instance of inappropriate error handling in matplotlib.

I have been able to reproduce this running python 3.5 and matplotlib 2.1.1

The following code can reproduce the error and requires plt.close() to clean out the pipeline before more plotting can be done:
```python
import matplotlib.pyplot as plt
data = [1,2,3,4,5]
ticks = ['any string']
plt.plot(data)
plt.xticks(ticks)
plt.show() 
```
```python
AttributeError: 'NoneType' object has no attribute 'seq'
```
The error is set-up by assigning non int/floats to the first argument of xticks. 

In:
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/category.py

StrCategoryConverter.convert seems to exist for the handling of strings but it doesn't seem to be handling this particular scenario properly.

on line 44 if you add:
```python
if axis.unit_data is None:
     axis.unit_data = UnitData([0])
````
It's a quick fix that makes it work but probably not the best solution. Its also of note that it is not just strings but data types other than int and float in the first argument of xticks will produce the same error.

While 
```python
plt.close()
```
Can be used to clear out the errors in jupyter notebook but I think that the error handling still can be improved. 
 
An alternative fix could be to prevent xticks from assigning bad values like this in 
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/pyplot.py
on lines 1662-1667:
```python
elif len(args)==1:
        if not any(isinstance(x, int) | isinstance(x, float) for x in args[0]):
            raise TypeError('First argument must be type int or float')
        locs = ax.set_xticks(args[0])
        labels = ax.get_xticklabels()
        
elif len(args)==2:
        if not any(isinstance(x, int) | isinstance(x, float) for x in args[0]):
            raise TypeError('First argument must be type int or float')
        locs = ax.set_xticks(args[0])
        labels = ax.set_xticklabels(args[1], **kwargs)   
```

**Comment by jklymak** [comment]
Thanks a lot for the clear report.  Without looking at it terribly deeply, I think this is a bug in unit handling, and the `category` unit handler should never have been called.  

Nonetheless, the where the error originates from is pretty clear in the error message:

```
Traceback (most recent call last):
  File "tickissue.py", line 5, in <module>
    plt.xticks(ticks)
```

I think your fixes above will just cover over the actual problem (and we never fix fundamental things in pyplot - its just a convenience wrapper).

**Comment by Rosuav** [comment]
The exception itself looks like an internal problem, though. It's confusing to see an AttributeError when the real problem is TypeError regarding the tick array. Perhaps the best solution is to trap the issue and raise something more meaningful?

**Comment by jklymak** [comment]
If you use #9776 which locks out other Converters once the convert has been set, then you get:

```
pythonw testticks.py
Traceback (most recent call last):
  File "testticks.py", line 5, in <module>
    plt.xticks(ticks)
  File "/Users/jklymak/matplotlib/lib/matplotlib/pyplot.py", line 1650, in xticks
    locs = ax.set_xticks(args[0])
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_base.py", line 3068, in set_xticks
    ret = self.xaxis.set_ticks(ticks, minor=minor)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axis.py", line 1660, in set_ticks
    ticks = self.convert_units(ticks)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axis.py", line 1508, in convert_units
    ret = self.converter.convert(x, self.units, self)
  File "/Users/jklymak/matplotlib/lib/matplotlib/units.py", line 186, in convert
    return np.asarray(obj, dtype=np.float64)
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.6/site-packages/numpy/core/numeric.py", line 531, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: could not convert string to float: 'any string'
```

**Comment by story645** [comment]
So this bug might be fixed in #9774 or #9783, but those are waiting on @tacaswell

**Comment by gurkpet** [comment]
Ya, I think a better error message would be appropriate.  It took me awhile to figure out why I could no longer plot anything until I restarted the kernel.  It will subsequently tell you the error is in whatever plotting function you are using rather than xticks.  It could be that I am just to new to have seen what should have been obvious but it was quite frustrating at the time as I knew that some of my plots worked even though I was getting the same 
```python
AttributeError: 'NoneType' object has no attribute 'seq'
```
error long after first calling the bad xticks call.  I think it should raise an error rather than allowing it to proceed to assign those values to anything that would continue to mess with plotting.  But I am new and this is my first ever attempt at a contribution to a project so I am well aware I could be very wrong.

**Comment by jklymak** [comment]
I think its definitely a bug, and hence its labelled, so thanks a lot for the report!

The problem is that `plt.xticks(['boo'])` should be a perfectly legal tick specification because Matplotlib now supports plotting calls like:

```python

fig, ax = plt.subplots()
ax.plot(['boo', 'who', 'aardvark'], [1, 10, 3])
ax.set_xticks(['boo', 'aardvark'])
```
kind of works, though the order gets screwed up...

![figure_1](https://user-images.githubusercontent.com/1562854/34578480-0012432c-f13a-11e7-8681-81620d1e63d7.png)

Your problem was that:

```python
fig, ax = plt.subplots()
ax.plot([1, 3, 10], [1, 10, 3])
ax.set_xticks(['boo', 'aardvark'])
```
is mixing "units", and the mysterious error was the result.

**Comment by QuLogic** [comment]
This appears to be the same as #12990 which was fixed by #13005.
